### ISSUE TITLE
`unnecessary-lambda-assignment` should not be generated when assigning to a “genuine” variable

### ISSUE THREAD

**Issue by barisione** [pylint-dev/pylint#8641]
### Bug description

I like the `unnecessary-lambda-assignment` warning but there are cases where I think using a lambda is clearer. For instance if the variable that's being assigned is genuinely used as a variable to which other callables are assigned.

For instance, in this example `callback` is assigned a method, a `functools.partial` and a `lambda`:
```
# pylint: disable=missing-module-docstring,missing-class-docstring,missing-function-docstring
# pylint: disable=pointless-statement,too-few-public-methods,unused-variable

import functools
import sys
from typing import Callable


class MyClass:
    def func(self, something: bool, something_else: bool, and_another: bool):
        callback: Callable[[str], str | None] | None = None
        if something:
            callback = self._foo
        elif something_else:
            callback = functools.partial(print, file=sys.stderr)
        elif and_another:
            calback = lambda msg: "Hello world"

        [...]  # Do something with callback

    def _foo(self, msg: str) -> str:
        [...]  # Do something interesting
        return msg.upper()

```

I think that, if I defined `callback` with `def` the code would be less readable (and potentially more confusing to people coming from other languages).

### Command used

```shell
pylint example.py
```

### Pylint output

```shell
************* Module example
example.py:17:22: C3001: Lambda expression assigned to a variable. Define a function using the "def" keyword instead. (unnecessary-lambda-assignment)

------------------------------------------------------------------
Your code has been rated at 9.33/10 (previous run: 9.33/10, +0.00)
```
```


### Expected behavior

No warning in this case.

### Pylint version

```shell
pylint 2.17.3
astroid 2.15.4
Python 3.10.1 (main, Jan 14 2022, 08:13:39) [Clang 13.0.0 (clang-1300.0.29.30)]
```

### OS / Environment

macOS 12.6

### Additional dependencies

N/A

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. I'm slightly -1 on this one. I think you're right and defining it with ``def`` in this case does break the flow of the code a little. But if the lambda do not use local variables, it can be created outside of the class and look the same than the other if branches, i.e.: 

```python
def hello_world_always(msg):
    return "Hello world"

class MyClass:
    def func(self, something: bool, something_else: bool, and_another: bool):
        ...
        elif:
            callback = hello_world_always

```        
If there is a local variable it's going to look worst with a def but not THAT much worse:
```python
class MyClass:
    def func(self, something: bool, something_else: bool, and_another: bool):
        ...
        elif:
            def hello_another_world(msg):
                return f"Hello world {and_another}"

            callback = hello_another_world
```        
```python
class MyClass:
    def func(self, something: bool, something_else: bool, and_another: bool):
        ...
        elif:
            # One local variable
            callback = lambda msg: f"Hello world {and_another}"
```        
    
Checking all this will be making the check more complex and slower and I'm not sure it's worth it. Keeping open to gather more opinions from contributors/maintainers.

**Comment by bariod** [comment]
Wouldn't this also contradict [PEP 8](https://peps.python.org/pep-0008/#programming-recommendations)?

> Always use a def statement instead of an assignment statement that binds a lambda expression directly to an identifier
